google/cloud/storage/client.cc

```diff
@@ -33,6 +32,61 @@ Client::Client(ClientOptions options)
     : Client(std::shared_ptr<internal::RawClient>(
           new internal::CurlClient(std::move(options)))) {}
 
+ObjectMetadata Client::UploadFileImpl(
+    std::string const& file_name,
+    internal::InsertObjectStreamingRequest request) {
+  std::ifstream is(file_name);
+  if (not is.is_open()) {
+    std::string msg = __func__;
+    msg += ": cannot open source file ";
+    msg += file_name;
+    google::cloud::internal::RaiseRuntimeError(msg);
+  }
+  if (not request.HasOption<DisableMD5Hash>() and
+      not request.HasOption<MD5HashValue>()) {
+    // Open a separate stream to read the file, because once we hit EOF there
+    // is no guarantee we can rewind the file to the beginning.
+    std::ifstream is2(file_name);
+    MD5_CTX md5;
+    MD5_Init(&md5);
+    std::string buffer;
+    buffer.resize(raw_client_->client_options().upload_buffer_size(), '\0');
+    while (not is2.eof()) {
+      is2.read(&buffer[0], buffer.size());
+      MD5_Update(&md5, &buffer[0], static_cast<std::size_t>(is2.gcount()));
+    }
+
+    std::string hash(MD5_DIGEST_LENGTH, ' ');
+    MD5_Final(reinterpret_cast<unsigned char *>(&hash[0]), &md5);
+    request.set_option(
+        MD5HashValue(internal::OpenSslUtils::Base64Encode(hash)));
+  }
+
+  auto result = raw_client_->WriteObject(request);
+  auto streambuf = std::move(result.second);
+
+  std::string buffer;
+  buffer.resize(raw_client_->client_options().upload_buffer_size(), '\0');
+  while (not is.eof() and is.good()) {
+    is.read(&buffer[0], buffer.size());
```

Comment:
Edge case: If the file is a stream/named pipe/something that may only be read from once, this will hang forever.  Might be worth adding a check at the beginning and failing if hashing is not disabled and the file is a FIFO.

Reply:
"File is a FIFO" is not the only condition under which this would fail.  You cannot rewind some devices (TTYs), but you can rewind others (`/dev/dsk`). Also "FIFO" is a Unix-ism, you cannot ask that question on Windows, you can ask "Is this a regular file?".  On Windows they have Named Pipes, which more or less do the same thing as a FIFO, but I do not think you can `_stat()` them, and then I do not know if you can query what they are. Furthermore, some devices never finish providing data (`/dev/zero`, `/dev/random`), so this would not work for them either.

Which is to say, trying to figure out under what conditions this works, and under what conditions it does not work is going to be hopeless.

I think the "Right Thing"[tm] is to add a comment that says "Use this with regular files only", so I did that.

Reply:
From looking at [the wiki](https://en.wikipedia.org/wiki/Named_pipe), it seems named pipes can't be mounted on a regular file system in Windows, so I think we should be able to call stat on the file at the given path and see if it's a regular file or not.  Looks like at least one of the other client libraries calls stat on a passed-in filename:
https://github.com/googleapis/google-cloud-python/blob/910846426428f24ad49e1a0f716e5bf103b3e3af/storage/google/cloud/storage/blob.py#L1088

As for what to do after calling stat, I really want to make it hard for users to shoot themselves in the foot, but also don't want to be overly restrictive.  How would you feel about checking if the file is a regular file, and if it's not, emitting a warning-level log statement?  That would at least make it easier for someone to figure out why their program is hanging indefinitely / behaving strangely.

Reply:
Done.